File: javascript/rest-api/rest-api.js
export class RestAPI {
  constructor(database = { users: [] }) {
    // Deep copy to avoid modifying the original seed data
    this.users = JSON.parse(JSON.stringify(database.users));
  }

  // Helper to find a user by name
  _findUser(name) {
    return this.users.find(user => user.name === name);
  }

  // Helper to update debt records, removing entries if amount is 0
  _updateDebt(debtor, creditor, amount) {
    if (amount === 0) {
      delete debtor.owes[creditor.name];
      delete creditor.owed_by[debtor.name];
    } else {
      debtor.owes[creditor.name] = amount;
      creditor.owed_by[debtor.name] = amount;
    }
  }

  get(url) {
    const urlParts = url.split('?');
    const path = urlParts[0];
    const params = new URLSearchParams(urlParts[1] || '');

    if (path === '/users') {
      const userNames = params.get('users');
      if (userNames) {
        const namesList = userNames.split(',');
        const filteredUsers = this.users.filter(user => namesList.includes(user.name));
        return { users: filteredUsers };
      } else {
        // Return a deep copy to prevent external modification
        return { users: JSON.parse(JSON.stringify(this.users)) };
      }
    }
    // Handle other GET endpoints if needed, though not specified by tests
    throw new Error(`Unknown GET endpoint: ${path}`);
  }

  post(url, payload) {
    if (url === '/add') {
      const userName = payload.user;
      // Check if user already exists (optional, but good practice)
      if (this._findUser(userName)) {
         throw new Error(`User ${userName} already exists.`);
      }
      const newUser = {
        name: userName,
        owes: {},
        owed_by: {},
        balance: 0,
      };
      this.users.push(newUser);
      // Return a copy
      return JSON.parse(JSON.stringify(newUser));
    }

    if (url === '/iou') {
      const { lender: lenderName, borrower: borrowerName, amount } = payload;
      const lender = this._findUser(lenderName);
      const borrower = this._findUser(borrowerName);

      if (!lender || !borrower) {
        throw new Error('Lender or borrower not found.');
      }

      // Calculate existing debt/credit between them
      const lenderOwesBorrower = lender.owes[borrowerName] || 0;
      const borrowerOwesLender = borrower.owes[lenderName] || 0; // Should be same as lender.owed_by[borrowerName]

      // Adjust balances first
      lender.balance += amount;
      borrower.balance -= amount;

      // Settle debts
      if (lenderOwesBorrower > 0) {
          if (lenderOwesBorrower >= amount) {
              // New loan is less than or equal to existing debt lender->borrower
              const remainingDebt = lenderOwesBorrower - amount;
              this._updateDebt(lender, borrower, remainingDebt);
          } else {
              // New loan is more than existing debt lender->borrower
              // Lender clears debt to borrower, and borrower now owes lender the difference
              const newDebtToLender = amount - lenderOwesBorrower;
              this._updateDebt(lender, borrower, 0); // Clear lender->borrower debt
              this._updateDebt(borrower, lender, newDebtToLender); // Create borrower->lender debt
          }
      } else {
          // No existing debt from lender to borrower, or borrower already owes lender
          const totalDebtToLender = borrowerOwesLender + amount;
          this._updateDebt(borrower, lender, totalDebtToLender);
      }


      // Return the updated lender and borrower, sorted alphabetically
      const updatedUsers = [lender, borrower]
        .sort((a, b) => a.name.localeCompare(b.name))
        .map(user => JSON.parse(JSON.stringify(user))); // Return copies

      return { users: updatedUsers };
    }

    // Handle other POST endpoints if needed
    throw new Error(`Unknown POST endpoint: ${url}`);
  }
}

File: python/rest-api/rest_api.py
import json

class RestAPI:
    def __init__(self, database=None):
        if database is None:
            self.database = {"users": []}
        else:
            self.database = database

    def get(self, url, payload=None):
        if url == "/users":
            if payload:
                data = json.loads(payload)
                names = data.get("users", [])
                users = [user for user in self.database["users"] if user["name"] in names]
            else:
                users = self.database["users"]
            # sort users by name
            users_sorted = sorted(users, key=lambda u: u["name"])
            return json.dumps({"users": users_sorted})
        else:
            return json.dumps({})

    def post(self, url, payload=None):
        data = json.loads(payload) if payload else {}
        if url == "/add":
            name = data["user"]
            new_user = {"name": name, "owes": {}, "owed_by": {}, "balance": 0.0}
            self.database["users"].append(new_user)
            return json.dumps(new_user)
        elif url == "/iou":
            lender_name = data["lender"]
            borrower_name = data["borrower"]
            amount = data["amount"]

            lender = next(user for user in self.database["users"] if user["name"] == lender_name)
            borrower = next(user for user in self.database["users"] if user["name"] == borrower_name)

            # Check if lender owes borrower
            if borrower_name in lender["owes"]:
                owed_amount = lender["owes"][borrower_name]
                if owed_amount > amount:
                    lender["owes"][borrower_name] -= amount
                    borrower["owed_by"][lender_name] -= amount
                    amount = 0
                elif owed_amount < amount:
                    amount -= owed_amount
                    del lender["owes"][borrower_name]
                    del borrower["owed_by"][lender_name]
                else:  # owed_amount == amount
                    del lender["owes"][borrower_name]
                    del borrower["owed_by"][lender_name]
                    amount = 0

            # If remaining amount > 0, update lender's owed_by and borrower's owes
            if amount > 0:
                lender["owed_by"].setdefault(borrower_name, 0.0)
                lender["owed_by"][borrower_name] += amount
                borrower["owes"].setdefault(lender_name, 0.0)
                borrower["owes"][lender_name] += amount

            # Update balances
            lender["balance"] = sum(lender["owed_by"].values()) - sum(lender["owes"].values())
            borrower["balance"] = sum(borrower["owed_by"].values()) - sum(borrower["owes"].values())

            # Prepare sorted user list
            users_sorted = sorted([lender, borrower], key=lambda u: u["name"])
            return json.dumps({"users": users_sorted})
        else:
            return json.dumps({})
